zrml/neo-swaps/src/migration.rs

```diff
@@ -0,0 +1,249 @@
+            // This should never fail; if it does, then we just delete the entry.
+            let mut liquidity_tree =
+                LiquidityTree::new(solo.owner.clone(), solo.total_shares).ok()?;
+            liquidity_tree.nodes.get_mut(0)?.fees = solo.fees; // Can't fail.
+            translated.saturating_inc();
+            Some(Pool {
+                account_id: pool.account_id,
+                reserves: pool.reserves,
+                collateral: pool.collateral,
+                liquidity_parameter: pool.liquidity_parameter,
+                liquidity_shares_manager: liquidity_tree,
+                swap_fee: pool.swap_fee,
+            })
+        });
+        log::info!("MigrateToLiquidityTree: Upgraded {} pools.", translated);
+        total_weight =
+            total_weight.saturating_add(T::DbWeight::get().reads_writes(translated, translated));
+        StorageVersion::new(NEO_SWAPS_NEXT_STORAGE_VERSION).put::<Pallet<T>>();
+        total_weight = total_weight.saturating_add(T::DbWeight::get().writes(1));
+        log::info!("MigrateToLiquidityTree: Done!");
+        total_weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        let old_pools =
+            storage_key_iter::<MarketIdOf<T>, OldPoolOf<T>, Twox64Concat>(NEO_SWAPS, POOLS)
+                .collect::<BTreeMap<_, _>>();
+        Ok(old_pools.encode())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(previous_state: Vec<u8>) -> Result<(), &'static str> {
+        let old_pools: BTreeMap<MarketIdOf<T>, OldPoolOf<T>> =
+            Decode::decode(&mut &previous_state[..])
+                .map_err(|_| "Failed to decode state: Invalid state")?;
+        let new_pool_count = Pools::<T>::iter().count();
+        assert_eq!(old_pools.len(), new_pool_count);
+        for (market_id, new_pool) in Pools::<T>::iter() {
+            let old_pool =
+                old_pools.get(&market_id).expect(&format!("Pool {:?} not found", market_id)[..]);
+            assert_eq!(new_pool.account_id, old_pool.account_id);
+            assert_eq!(new_pool.reserves, old_pool.reserves);
+            assert_eq!(new_pool.collateral, old_pool.collateral);
+            assert_eq!(new_pool.liquidity_parameter, old_pool.liquidity_parameter);
+            assert_eq!(
+                new_pool.liquidity_shares_manager.nodes,
+                vec![Node {
+                    account: Some(old_pool.account_id.clone()),
+                    stake: old_pool.liquidity_shares_manager.total_shares,
+                    fees: old_pool.liquidity_shares_manager.fees,
+                    descendant_stake: Zero::zero(),
+                    lazy_fees: Zero::zero(),
+                }]
+            );
+            assert_eq!(new_pool.swap_fee, old_pool.swap_fee);
+        }
+        log::info!("MigrateToLiquidityTree: Post-upgrade pool count is {}!", new_pool_count);
+        Ok(())
+    }
```

Comment:
Already tested this out?

Reply:
Yeah, but there was a mistake in the test, it's fixed now. 8bfdfc0ce484665d2627d97826b2fa72eff657ad